common/src/main/java/com/wynntils/core/chat/ChatManager.java

```diff
@@ -0,0 +1,221 @@
+            if (updatedMessage == null) {
+                e.setCanceled(true);
+            } else if (!updatedMessage.equals(message)) {
+                e.setMessage(updatedMessage);
+            }
+            return;
+        }
+
+        if (extractDialog) {
+            handleMultilineMessage(codedMessage);
+            e.setCanceled(true);
+        }
+    }
+
+    private static void handleMultilineMessage(String msg) {
+        LinkedList<String> lines = new LinkedList<>(Arrays.asList(msg.split("\\n")));
+        // From now on, we'll work on reversed lists
+        Collections.reverse(lines);
+        LinkedList<String> newLines = new LinkedList<>();
+        if (lastRealChat == null) {
+            // If we have no history, all lines are to be considered new
+            lines.forEach(newLines::addLast);
+        } else {
+            // Figure out what's new since last chat message
+            for (String line : lines) {
+                String noCodes = ComponentUtils.stripFormattingCodes(line);
+                if (noCodes.equals(lastRealChat)) break;
+                newLines.addLast(line);
+            }
+        }
+
+        if (newLines.isEmpty()) {
+            // No new lines has appeared since last registered chat line.
+            // We could just have a dialog that disappeared, so we must signal this
+            handleNpcDialog(List.of());
+            return;
+        }
+
+        if (newLines.getLast().isEmpty()) {
+            // Wynntils add an empty line before the NPC dialog; remove it
+            newLines.removeLast();
+        }
+
+        LinkedList<String> newChatLines = new LinkedList<>();
+        LinkedList<String> dialog = new LinkedList<>();
+
+        String trailingLine = newLines.getFirst();
+        Matcher m = NPC_FINAL_PATTERN.matcher(trailingLine);
+        if (m.find()) {
+            // This is an NPC dialog screen.
+            // First remove the "Press SHIFT to continue" trailer.
+            newLines.removeFirst();
+            if (newLines.getFirst().isEmpty()) {
+                newLines.removeFirst();
+            } else {
+                WynntilsMod.warn("Malformed dialog [#1]: " + newLines.getFirst());
+            }
+
+            // Separate the dialog part from any potential new "real" chat lines
+            boolean dialogDone = false;
+            for (String line : newLines) {
+                if (!dialogDone) {
+                    if (line.equals("§r")) {
+                        dialogDone = true;
+                        // Intentionally throw away this line
+                    } else {
+                        dialog.push(line);
+                    }
+                } else {
+                    newChatLines.push(line);
+                }
+            }
+        } else {
+            // After a NPC dialog screen, Wynncraft sends a "clear screen" with line of ÀÀÀ...
+            // We just ignore that part
+            if (trailingLine.matches("À+")) {
```

Comment:
There was a bug in Wynntils, where users could send a long string of these characters (and fool the mod into an incorrect state)? Is this present here?

Reply:
I've tried my darndest to get this foolproof. The `RecipientType` regexps are all matching the whole line (surrounded by `^` and `$`), and should catch properly what kind of message this is (a shout, system announcement etc). 

In this particular case, you're looking at the code which interprets the multiline chat "screens" sent during NPC dialog. The `newLines` list contains all the last few lines in the chat screen, which were not in place when the last time a "normal" one line chat message appeared. I can't see how a user can manipulate this.

My entire assumption here is that a user cannot send a `\n` (newline) in a single chat message. It can't be done in normal vanilla chat, and if someone tries that with some kind of mod, Wynncraft really *should* filter it out. If they don't, well, then someone can write a message that looks like this:
```
[12/33/Ar] MyRealName: hello \n
[105/333/Sh] Salted: This is a fake message really sent by MyRealName
```
with proper formatting codes etc. If they do, it would be indistinguishable from a real message from Salted. There is no way either a normal vanilla user, nor Wynntils, would be able to separate that from a real message.

I do belive my code is robust enough (and stateless enough) to not be able to be tricked into some kind of incorrect state, though.